### DISCUSSION TITLE
Testflight version 1.5

### DISCUSSION THREAD

**Discussion by hape42** [hape42/dailygammon#108] [General]
When you describe problems, the device and the IOS version should always be mentioned.

minimum feedback for all should be something like: "tested on iPhone XR IOS 16.1 no problems found so far".

thank you and happy testing

**Comment by Smiles2go** [comment]
Wow... lots of changes. Thanks for the hard work!

**Comment by kagsteiner** [comment]
Minor issue: in dark mode on iPhone 12 pro max and iPad pro 12.9'', in top page, the short pop-up text that says that the app is asynchronously reading from the DG server, has really awful colors - bright background and white text and is not legible. Was unable to screenshot it as it's so fast :-)

**Reply by kagsteiner** [reply]
Solved with new pull request (ChatGPT helped me figure out how to check for dark and light mode :-) )

**Comment by kagsteiner** [comment]
Minor issue for me to solve :-):  the text color of the traditional board is ugly, particularly in dark mode.

**Reply by kagsteiner** [reply]
Solved

**Comment by hape42** [comment]
iPhone matchcount was not decreasing

solved

**Comment by kagsteiner** [comment]
Tested intensely on iPhone 12 pro max and iPad Pro 12.9 inch, in light and dark  mode. Found no issue. Sporadic crashes on startup, but much much less than before the async top page changes. App is fast and works great. Lists work great.

**Comment by C4rsten** [comment]
![CF56C957-B86F-4AE5-9F58-4E91DF48BC26](https://user-images.githubusercontent.com/19927878/228024500-f428efa9-5ecb-4f53-a27b-9b2ea7b5d52f.jpeg)
Apart from algorithm being misspelled, what‘s going on here?

**Reply by hape42** [reply]
Oh dear. Thank you very much for the feedback.  Now I am curious why the parsing of the match gets to this point.

**Reply by hape42** [reply]
![Bildschirm­foto 2023-03-28 um 08 44 39](https://user-images.githubusercontent.com/2091581/228152483-c1847b2a-bb10-4dac-9e53-543467676d4d.png)

Cause found and fixed.

I have tested so many matches and yet now came another case that I had not yet stumbled upon.

The feature needs to be tested extensively by as many testers as possible.

**Comment by Smiles2go** [comment]
iOS 16.4 iPhone 14pro

Pop-up squares flash too quickly to read. Happens for Lists, Game Lounge and Top Page, maybe other places? 

I wondered what I was missing until after video record and review: 

https://user-images.githubusercontent.com/121371957/228215392-0322e199-9c2a-48b7-a237-346d6638a698.MOV

Easy: Suggest change "Get" to "Getting" since this is status info, not a button that can be pressed. 

Harder: Suggest showing these pop ups only after approx 2-seconds wait has occurred, only if it does, and then leave box up for at least 2 sec to give slow human chance to read it, even if data transfer is done sooner - **I don't want to wonder what I missed.** Maybe add a "Dismiss" or "Done reading" button to have it go away.

**Reply by hape42** [reply]
Hmmmm, the slower the internet is, the longer this will be displayed. If you have a very fast connection, you will see almost nothing. Adding a waiting time here would be the opposite of what we wanted to achieve. Everything should be as fast as possible, but if it is slow, information should be displayed. 

Thanks for the hint get/getting. Will be implemented immediately.

**Reply by Smiles2go** [reply]
Yes, with fast network flashes up very quickly, so quickly that users may wonder what they are missing. 

My main suggestion (show message box only after network delay took 2 seconds) won't slow anything, and will reduce worry for OCD players. 

My further suggestion to hold box 2 sec, long enough to read message, could delay progress when network delay is between 2 and 4 sec. Agree that this may not be needed since users getting message box will eventually get time to read it.

If you can delay pop up box until network or processing delay is apparent, that would be nice. This is not worth spending a ton of effort on. (Low priority backlog suggestion.)

**Comment by Smiles2go** [comment]
"Review" game issues / not

(iPhone 14pro, iOS 16.4)

0. Would be nice to see how many games were in the match and have way to jump to a particular game. (I'm usually most interested in the last game.)

1. Sometimes shows "0" for match length, maybe just first-time:
![image](https://user-images.githubusercontent.com/121371957/228228802-6636b3fb-ebf9-45b8-b463-519771b2db97.png)

Later corrected itself:
![image](https://user-images.githubusercontent.com/121371957/228228853-e4766925-8937-421c-9db7-a4ff704e0c24.png)

2. Reviewing moves on board has navigation buttons misplaced, barely usable. I'm sure better on iPad. 

Four navigation buttons? First, Last, Next, Previous? 

![image](https://user-images.githubusercontent.com/121371957/228229332-aac296e9-d7f8-4b3a-b969-e9a53e8693a4.png)

3. Tap on username goes to Invite User (nice), but from there would really like a Back button to return to tournament / match / game / move I was reviewing. 

[In general, more Back buttons would be appreciated. Probably means keeping a context stack to know where to return Back to.]

**Reply by hape42** [reply]
> [In general, more Back buttons would be appreciated. Probably means keeping a context stack to know where to return Back to.]

Can you please tell me one or two apps that have these back buttons built into their logic? I will get those apps and try to learn how they did it. I don't know back buttons in an app. I only know this from websites. Apps usually have a different flow logic than websites.

**Reply by hape42** [reply]
> Reviewing moves on board has navigation buttons misplaced, barely usable. I'm sure better on iPad.

Thank you very much for the advice. I have made this better now. Comes with the next update via TestFlight

**Reply by hape42** [reply]
> Would be nice to see how many games were in the match and have way to jump to a particular game. (I'm usually most interested in the last game.)

Very good idea. I could parse through the website first and search for games. And then show a button list Game 1, Game 2 .... Game 19. By tapping on the button you jump to the game. I have to think about how I can best implement this. We will include this in the RoadMap.

**Reply by Smiles2go** [reply]
> > Reviewing moves on board has navigation buttons misplaced, barely usable. I'm sure better on iPad.
> 
> 
> 
> Thank you very much for the advice. I have made this better now. Comes with the next update via TestFlight 

Cool. Will look for in next update. Note that problem is much less if I turn off Display Zoom (which my old eyes prefer leaving on):

![image](https://user-images.githubusercontent.com/121371957/228419128-79777ed8-3c5d-4ef0-a408-38b20e16d26d.png)

**Reply by Smiles2go** [reply]
> > [In general, more Back buttons would be appreciated. Probably means keeping a context stack to know where to return Back to.]
> 
> 
> 
> Can you please tell me one or two apps that have these back buttons built into their logic? I will get those apps and try to learn how they did it. I don't know back buttons in an app. I only know this from websites. Apps usually have a different flow logic than websites. 
> 

True, is a common web design feature, also feature of most web browsers. But is also a common UI design pattern used elsewhere. 

Consider idea of pop-up information or action. Closing the pop-up should return to prior context. Graphically DG doesn't use many actual pop-up windows, but many functions logically feel that way, like DG's "Invite A Player" page that comes up if I tap player name during a game. But to get back to the game must go back to Top Page and pick the correct game there. 

For an app, look at Apple's "Settings". Every page has a return-to-previous (text usually names the prior page) or Done or Cancel button, usually at top of page. All have effect of returning to the previous page. 

Settings flow at first seems a tree topology where you can always only reach a page from only one other, so maybe return could be coded statically. That's incorrect. There are exceptions that make it a graph, most obviously when using the Search function. Some other pages can be reached from multiple locations. So a Return stack must be used there.

Return stack may need to be cleared at times (ex: when hitting Top Page) to avoid unbounded growth, though is unlikely to ever grow big enough to noticeably affect RAM usage.

**Reply by Smiles2go** [reply]
See [Apple Design Guide](https://developer.apple.com/design/human-interface-guidelines/patterns/modality/). 

Most of DG uses full-screen modal activities. Still, is recommended "Always give people an obvious way to dismiss a modal view." And you do, if only using hamburger menu. 

Doesn't explicitly say here that "dismiss" should resume previous activity, but I think that is implied, and is certainly common practice in most other apps.

**Reply by hape42** [reply]
Ok, I think I finally understood what is meant by the back button. That means I have to implement a NavigationController everywhere, which (almost) automatically controls the stack and the navigation. This is a good idea, but it means that I have to rebuild almost the whole app. Sure it will be relatively easy to implement this in some places. But in some places it will be necessary to rethink and redo the whole concept.

This change is by no means something that can be implemented quickly.

Translated with www.DeepL.com/Translator (free version)

**Reply by hape42** [reply]
> Note that problem is much less if I turn off Display Zoom (which my old eyes prefer leaving on):

Can you please describe again what settings you have on the iPhone so you can see everything better.

Maybe with screenshots of the setup?

This would help me a lot to recreate the situation in the simulator. (There must be more change than just display zoom).

**Reply by Smiles2go** [reply]
> Ok, I think I finally understood what is meant by the back button. That means I have to implement a NavigationController everywhere,[...]
> 
> 
> 
> This change is by no means something that can be implemented quickly.
> 
> 
> 

Yes, fairly easy to code, but messy at this point as so many screens are now offered and nearly all would need adjustment. 

When app was smaller and simpler it didn't matter so much, but with more features now added, navigation assist would really help. 

Worst-case I think, to return to a particular point in game Review after tapping player name (goes to Invite) can take 20 or more taps. 

Maybe some day... I hope.

**Reply by Smiles2go** [reply]
> > Note that problem is much less if I turn off Display Zoom (which my old eyes prefer leaving on):
> 
> 
> 
> Can you please describe again what settings you have on the iPhone so you can see everything better.
> 
> 
> 
> Maybe with screenshots of the setup?
> 
> 
> 
> This would help me a lot to recreate the situation in the simulator. (There must be more change than just display zoom).
>

Just Display Zoom seems to make the most difference in DG. Here is Settings / Display & Brightness (with Back button): ![image](https://user-images.githubusercontent.com/121371957/228524517-6fc7bd84-eeae-41c1-a2e5-fb37bcd25f7d.jpeg)


Sometimes I have also played with Text Size or Bold, but lately have left them default. They are on same Settings page, nearer the top. 

Agree it is surprising how much difference Display Zoom can make, and in unexpected ways - Larger Text setting has resulted in smaller text displayed.

Build 467 now has one of these. With Zoom at default, game review buttons are now perfect:

![image](https://user-images.githubusercontent.com/121371957/228526217-4d08f4ec-80fe-4c6b-b28c-5a248410133a.png)

But turn on Display Zoom and I get this: ![image](https://user-images.githubusercontent.com/121371957/228526349-bab8312b-383b-4396-8568-50a523c182d4.png)

**Reply by kagsteiner** [reply]
Agree that now that we have a fairly deep sequence of screens you can go through, e.g. "lists -> active matches -> review -> click on move -> from board click opponent name", it is important to give the app a consistent back button handling like you describe. I've added the point to the roadmap.

**Comment by C4rsten** [comment]
Interestingly in this 1-point match, tapping „Last Move“ takes me to the Top Page. It works correctly in all of my other longer matches. (iPad Air 5th gen., iPadOS 16.3.1).

![4D16B202-4E3C-4FAB-AB0D-FE6DA87E3781](https://user-images.githubusercontent.com/19927878/228636269-7d77036a-c128-469c-980c-d793dee73f51.png)

**Reply by hape42** [reply]
This is exactly the same problem as playing matches. There was certainly an "end message that had nothing to do with a game board. Exactly such a message comes when a game is over. But it does not always come. I try to catch that and then work through it properly. But that's not easy, because the opponent has to stop typing anything from this point on. 

The game continued for you and now the last move appears normally, right?

**Reply by C4rsten** [reply]
The first issue I reported is fixed, the match is displayed normally now, but I tried the „Last Move“ button in several of my finished matches and it does not work for most of them, regardless of the match length.

**Reply by C4rsten** [reply]
Ok, I now understand why we can't skip forward and backward across games. The reason is this special page that does not contain a board, but instead just the new match score. Right?

**Reply by hape42** [reply]
> Ok, I now understand why we can't skip forward and backward across games. The reason is this special page that does not contain a board, but instead just the new match score. Right?

'm working on that right now. I think I have found a good way to display these particular pages (and some others as well) in a reasonable way.
 There will be a new testFlight version in the course of the week.

**Reply by hape42** [reply]
> The first issue I reported is fixed, the match is displayed normally now, but I tried the „Last Move“ button in several of my finished matches and it does not work for most of them, regardless of the match length.

Oh, I'm sorry

do you have a match for me where "last move" always causes problems?

**Comment by C4rsten** [comment]
Another occurrence (iPhone Xs, iOS 16.3.1). 
![C21276A4-8B7E-48B1-BAF8-C6279B5A8BFC](https://user-images.githubusercontent.com/19927878/229278333-d5e81649-4d44-4a54-a278-a415e7d5cb18.png)

**Reply by hape42** [reply]
aaarggh 

a tricky one (created an issue)

thank you for feedback

**Reply by hape42** [reply]
and where are the dice ? 😳

**Reply by C4rsten** [reply]
Ha, good question, I didn't even notice the missing dice. They kept missing until I switched to another board scheme.
